File: challenges/2026-09-25-relay-hop-budget/solution.java
import java.util.*;

public class solution {

    public static int minRelayCost(int n, List<int[]> edges, int src, int dst, int maxHops) {
        // TODO: implement
        return 0;
    }

    public static void main(String[] args) {
        List<int[]> edges = new ArrayList<>();
        edges.add(new int[]{0, 1, 100});
        edges.add(new int[]{1, 2, 100});
        edges.add(new int[]{0, 2, 500});
        edges.add(new int[]{2, 3, 100});
        edges.add(new int[]{1, 3, 300});

        List<Object[]> tests = new ArrayList<>();
        tests.add(new Object[]{"example 1", 4, edges, 0, 3, 2, 400});
        tests.add(new Object[]{"example 2", 4, edges, 0, 3, 1, -1});
        tests.add(new Object[]{"example 3", 4, edges, 0, 3, 3, 300});
        tests.add(new Object[]{"edge: zero hop budget", 4, edges, 0, 3, 0, -1});
        tests.add(new Object[]{"edge: unreachable dst", 4, List.of(new int[]{0, 1, 5}), 0, 3, 3, -1});
        tests.add(new Object[]{"edge: cheaper duplicate direct link wins", 2,
                List.of(new int[]{0, 1, 7}, new int[]{0, 1, 3}), 0, 1, 1, 3});

        for (Object[] test : tests) {
            String name = (String) test[0];
            int n = (Integer) test[1];
            @SuppressWarnings("unchecked")
            List<int[]> edgeList = (List<int[]>) test[2];
            int src = (Integer) test[3];
            int dst = (Integer) test[4];
            int maxHops = (Integer) test[5];
            int expected = (Integer) test[6];

            int got = minRelayCost(n, edgeList, src, dst, maxHops);
            String status = got == expected ? "PASS" : "FAIL";
            System.out.println("[" + status + "] " + name + ": expected=" + expected + " actual=" + got);
        }
    }
}

File: challenges/2026-09-25-relay-hop-budget/solution.go
package main

import "fmt"

func minRelayCost(n int, edges [][]int, src int, dst int, maxHops int) int {
	// TODO: implement
	return 0
}

type testCase struct {
	name     string
	n        int
	edges    [][]int
	src      int
	dst      int
	maxHops  int
	expected int
}

func main() {
	edges := [][]int{{0, 1, 100}, {1, 2, 100}, {0, 2, 500}, {2, 3, 100}, {1, 3, 300}}

	tests := []testCase{
		{"example 1", 4, edges, 0, 3, 2, 400},
		{"example 2", 4, edges, 0, 3, 1, -1},
		{"example 3", 4, edges, 0, 3, 3, 300},
		{"edge: zero hop budget", 4, edges, 0, 3, 0, -1},
		{"edge: unreachable dst", 4, [][]int{{0, 1, 5}}, 0, 3, 3, -1},
		{"edge: cheaper duplicate direct link wins", 2, [][]int{{0, 1, 7}, {0, 1, 3}}, 0, 1, 1, 3},
	}

	for _, tc := range tests {
		got := minRelayCost(tc.n, tc.edges, tc.src, tc.dst, tc.maxHops)
		status := "FAIL"
		if got == tc.expected {
			status = "PASS"
		}
		fmt.Printf("[%s] %s: expected=%v actual=%v\n", status, tc.name, tc.expected, got)
	}
}
